⭐ raw, position and sensor data gdp drifter dataset adapter

kevinsantana11 opened the pull request:
starts #377

selipot commented:
review now or eventually?

kevinsantana11 commented:
> review now or eventually?

Just a place holder PR for now. Once it's ready for review I'll send another comment mentioning you.

selipot commented:
I tried
```
from clouddrift.adapters.gdp import rawfiles as raw
ra = raw.to_raggedarray()
```

which led to the following outputs and errors:
```
https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/dirfl_1_5000.dat: 100%|██████████| 586k/586k [00:00<00:00, 1.04MB/s]
https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/dirfl_5001_10000.dat: 100%|██████████| 586k/586k [00:00<00:00, 911kB/s] 
https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/dirfl_10001_15000.dat: 100%|██████████| 620k/620k [00:02<00:00, 290kB/s] 
https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/dirfl_15001_20000.dat: 66.2kB [00:00, 205kB/s]
https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/dirfl_15001_current.dat: 100%|██████████| 1.56M/1.56M [00:00<00:00, 1.66MB/s]
https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/buoydata_1_5000_rawfiles.dat.gz: 100%|██████████| 247M/247M [01:49<00:00, 2.36MB/s] 
loading data files:   0%|          | 0/1 [00:00<?, ?it/s]/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:212: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  cur_df = pd.read_csv(
loading data files: 100%|██████████| 1/1 [00:32<00:00, 32.67s/it]
Retrieving the number of obs: 100%|███████| 27892/27892 [08:03<00:00, 57.72it/s]
[/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:407](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:407): UserWarning: Variable ID requested but not found; skipping.
  warnings.warn(f"Variable {var} requested but not found; skipping.")
[...]
[/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440): UserWarning: Variable End_date requested but not found; skipping.
  warnings.warn(
[/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440): UserWarning: Variable End_lat requested but not found; skipping.
  warnings.warn(
[/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440): UserWarning: Variable End_lon requested but not found; skipping.
  warnings.warn(
[/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/clouddrift/raggedarray.py:440): UserWarning: Variable Drogue_off_date requested but not found; skipping.
  warnings.warn(
Filling the Ragged Array:   0%|              | 67/27892 [00:03<25:28, 18.21it/s]
```
and the error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File [/Users/selipot/projects.git/clouddrift/create_ragged.py:3](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/create_ragged.py:3)
      [1](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/create_ragged.py:1) # %% try the ragged array generation of raw files
      [2](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/create_ragged.py:2) from clouddrift.adapters.gdp import rawfiles as raw
----> [3](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/create_ragged.py:3) ra = raw.to_raggedarray()

File [~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:231](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:231), in to_raggedarray(kind, tmp_path)
    [227](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:227)     else:
    [228](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:228)         df = pd.concat([df, cur_df])
--> [231](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:231) ra = RaggedArray.from_files(
    [232](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:232)     indices=gdp_metadata_df[["ID"]].values,
    [233](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:233)     preprocess_func=setup_preprocess(gdp_metadata_df, df, config),
    [234](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:234)     name_coords=config.coords,
    [235](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:235)     name_meta=[
    [236](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:236)         "ID",
    [237](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:237)         "WMO_number",
    [238](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:238)         "program_number",
    [239](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:239)         "buoys_type",
    [240](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:240)         "Start_date",
    [241](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:241)         "Start_lat",
    [242](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:242)         "Start_lon",
    [243](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:243)         "End_date",
    [244](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:244)         "End_lat",
    [245](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:245)         "End_lon",
    [246](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:246)         "Drogue_off_date",
    [247](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:247)         "death_code",
    [248](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:248)     ],
    [249](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:249)     name_data=["lat", "lon", "drogue", "sst", "voltage", "sensor4", "sensor5", "sensor6"],
    [250](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:250)     name_dims={"traj": "rows", "obs": "obs"},
    [251](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:251)     rowsize_func=lambda id_: len(df[df["id"] == id_[0]])
    [252](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:252) )
    [253](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:253) return ra

File [~/projects.git/clouddrift/clouddrift/raggedarray.py:142](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:142), in RaggedArray.from_files(cls, indices, preprocess_func, name_coords, name_meta, name_data, name_dims, rowsize_func, attrs_global, attrs_variables, **kwargs)
    [136](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:136) rowsize_func = (
    [137](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:137)     rowsize_func
    [138](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:138)     if rowsize_func
    [139](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:139)     else lambda i, **kwargs: preprocess_func(i, **kwargs).sizes["obs"]
    [140](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:140) )
    [141](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:141) rowsize = cls.number_of_observations(rowsize_func, indices, **kwargs)
--> [142](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:142) coords, metadata, data, coord_dims = cls.allocate(
    [143](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:143)     preprocess_func,
    [144](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:144)     indices,
    [145](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:145)     rowsize,
    [146](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:146)     name_coords,
    [147](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:147)     name_meta,
    [148](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:148)     name_data,
    [149](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:149)     name_dims,
    [150](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:150)     **kwargs,
    [151](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:151) )
    [153](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:153) extracted_attrs_global, extracted_attrs_variables = cls.attributes(
    [154](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:154)     preprocess_func(indices[0], **kwargs),
    [155](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:155)     name_coords,
    [156](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:156)     name_meta,
    [157](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:157)     name_data,
    [158](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:158) )
    [160](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:160) if attrs_global is None:

File [~/projects.git/clouddrift/clouddrift/raggedarray.py:425](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:425), in RaggedArray.allocate(preprocess_func, indices, rowsize, name_coords, name_meta, name_data, name_dims, **kwargs)
    [417](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:417) # loop and fill the ragged array
    [418](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:418) for i, index in tqdm(
    [419](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:419)     enumerate(indices),
    [420](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:420)     total=len(indices),
   (...)
    [423](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:423)     disable=_DISABLE_SHOW_PROGRESS,
    [424](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:424) ):
--> [425](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:425)     with preprocess_func(index, **kwargs) as ds:
    [426](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:426)         size = rowsize[i]
    [427](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/raggedarray.py:427)         oid = index_traj[i]

File [~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:180](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:180), in setup_preprocess.<locals>.wrapped_preprocess(id_)
    [164](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:164) traj_data_df = data_df[data_df["id"] == id_[0]]
    [166](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:166) variables = {
    [167](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:167)     "wmo_number": (["traj"], traj_md_df[["WMO_number"]].values[0]),
    [168](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:168)     "program_number": (["traj"], traj_md_df[["program_number"]].values[0]),
   (...)
    [177](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:177)     "death_code": (["traj"], traj_md_df[["death_code"]].values[0]),
    [178](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:178) }
--> [180](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:180) data_variables = config.get_vars_config_map("obs", traj_data_df)
    [181](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:181) coords = config.get_coords_config_map("obs", traj_data_df)
    [182](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:182) row_coord = {
    [183](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:183)     "id": (["traj"], traj_md_df[["ID"]].values[0]),
    [184](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:184) }

File [~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:51](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:51), in ParsingConfiguration.get_vars_config_map(self, dim_name, df)
     [50](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:50) def get_vars_config_map(self, dim_name: str, df: pd.DataFrame):
---> [51](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:51)     all_config_map = self._get_all_config_map(dim_name, df)
     [52](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:52)     return {col: all_config_map[col] for col in all_config_map.keys() if col not in self.coords}

File [~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:60](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:60), in ParsingConfiguration._get_all_config_map(self, dim_name, df)
     [58](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:58) def _get_all_config_map(self, dim_name: str, df: pd.DataFrame):
     [59](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:59)     pre_transform = {col: ([dim_name], df[[col]].to_numpy().flatten()) for col in self.cols}
---> [60](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:60)     post_transform = self._apply_transform(pre_transform)
     [61](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:61)     return post_transform

File [~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:70](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:70), in ParsingConfiguration._apply_transform(self, variable_config_map)
     [68](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:68)         _, variable = variable_config_map[col]
     [69](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:69)         args.append(variable)
---> [70](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:70)     variable_config_map[output_col] = ([dim], func(*args))
     [71](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:71)     [variable_config_map.pop(col) for col in input_cols]
     [72](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:72) return variable_config_map

File [~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:38](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:38), in _parse_datetime_with_day_ratio(monthSeries, day_series, year_series)
     [36](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:36)     dayratio = day_with_ratio - day
     [37](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:37)     seconds = dayratio * _SECONDS_IN_DAY
---> [38](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:38)     dt_ns = (datetime.datetime(year=int(year), month=int(month), day=int(day)) + datetime.timedelta(seconds=seconds)).timestamp() * 10**9
     [39](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:39)     values.append(int(dt_ns))
     [40](https://file+.vscode-resource.vscode-cdn.net/Users/selipot/projects.git/clouddrift/~/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:40) return np.array(values, dtype="datetime64[ns]")

ValueError: year 198048037 is out of range
```

philippemiron commented:
looks like it's taking the date as the year.

selipot commented:
I tested the following, it ran, but I am not sure of the output

```
In [1]: from clouddrift.datasets import gdpraw

In [2]: ds = gdpraw()

https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/buoydata_5001_10000_rawfiles.dat.gz: 100%|███████████████████████████████████████████████████████████████████████████████| 723M/723M [05:42<00:00, 2.22MB/s]
https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/buoydata_10001_15000_rawfiles.dat.gz: 100%|██████████████████████████████████████████████████████████████████████████████| 834M/834M [05:47<00:00, 2.51MB/s]
https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/buoydata_15001_current_rawfiles.dat.gz: 100%|██████████████████████████████████████████████████████████████████████████| 2.67G/2.67G [19:03<00:00, 2.51MB/s]
Cannot determine if the file has been updated on the remote source. 'Last-Modified' header not present in server response.
https://www.aoml.noaa.gov/ftp/pub/phod/buoydata/dirfl_15001_20000.dat: 65.6kB [00:00, 408kB/s]
Loading files:   0%|                                    | 0/4 [00:00<?, ?file/s]/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:103: RuntimeWarning: overflow encountered in cast
  post_remove[[col]].to_numpy().flatten().astype(self.col_dtypes[col])
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:437: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  for idx, chunk in enumerate(file_chunks):
Processing file chunks: 100%|██████████████| 198/198 [03:27<00:00,  1.05s/chunk]
Loading files:  50%|█████████████▌             | 2/4 [05:41<05:53, 176.93s/file]/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:377: UserWarning: Data has drifter ids not found in the metadata table. Please inspect the following ids████| 517/517 [01:23<00:00,  8.58chunk/s]
  warnings.warn(
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:377: UserWarning: Data has drifter ids not found in the metadata table. Please inspect the following ids
  warnings.warn(
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:377: UserWarning: Data has drifter ids not found in the metadata table. Please inspect the following ids
  warnings.warn(
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:377: UserWarning: Data has drifter ids not found in the metadata table. Please inspect the following ids
  warnings.warn(
/Users/selipot/projects.git/clouddrift/clouddrift/adapters/gdp/rawfiles.py:437: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  for idx, chunk in enumerate(file_chunks):
Processing file chunks: 100%|██████████████| 517/517 [03:31<00:00,  2.45chunk/s]
Processing file chunks:  99%|█████████████▉| 566/569 [09:05<00:02,  1.04chunk/s]
Loading files: 100%|███████████████████████████| 4/4 [21:04<00:00, 316.06s/file]
Processing file chunks: 100%|████████████| 2045/2045 [04:09<00:00,  8.20chunk/s]
merging drifter chunks: 100%|███████| 27842/27842 [03:36<00:00, 128.35drifter/s]

In [3]: ds
Out[3]: 
<xarray.Dataset> Size: 15GB
Dimensions:            (obs: 332274093, traj: 27842)
Coordinates:
    posObsDatetime     (obs) datetime64[ns] 3GB 1979-01-01T05:00:00 ... 2024-...
    sensorObsDatetime  (obs) datetime64[ns] 3GB 1996-08-07T17:52:30 ... 2024-...
    id                 (traj) int64 223kB 9320177 7711726 ... 300534063909050
Dimensions without coordinates: obs, traj
Data variables: (12/20)
    drogue             (obs) float16 665MB 0.0 0.0 0.0 0.0 ... 21.0 22.0 22.0
    lat                (obs) float32 1GB 0.0 0.0 0.0 ... -57.82 -57.82 -57.82
    lon                (obs) float32 1GB 0.0 0.0 0.0 0.0 ... 82.88 82.88 82.88
    sensor4            (obs) float32 1GB 0.0 0.0 0.0 ... 1.014e+03 1.013e+03
    sensor5            (obs) float32 1GB 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0
    sensor6            (obs) float32 1GB 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0
    ...                 ...
    program_number     (traj) int64 223kB 1325 1346 1325 ... 31412 31412 31412
    rowsize            (traj) int64 223kB 3558 2189 1318 249 ... 964 951 941 920
    start_date         (traj) datetime64[ns] 223kB 1996-08-08T15:05:00 ... 20...
    start_lat          (traj) float64 223kB 0.23 20.2 -0.53 ... -59.66 -59.94
    start_lon          (traj) float64 223kB 172.1 201.2 172.5 ... 84.94 80.08
    wmo_number         (traj) int64 223kB 5200604 5100554 ... 7801695 5802066
```

selipot commented:
I think the "posObsDatetime" and "sensorObsDatetime" may not necessarily match but in my case they really don't always match, see:
```
In [9]: ds["sensorObsDatetime"][0:10].values
Out[9]: 
array(['1996-08-07T17:52:30.000000000', '1996-08-07T19:28:07.500000000',
       '1996-08-07T21:54:22.500000000', '1996-08-07T23:35:37.500000000',
       '1996-08-08T04:56:15.000000000', '1996-08-08T06:37:30.000000000',
       '1996-08-08T08:13:07.500000000', '1996-08-08T10:33:45.000000000',
       '1996-08-08T12:09:22.500000000', '1996-08-08T17:41:15.000000000'],
      dtype='datetime64[ns]')

In [10]: ds["posObsDatetime"][0:10].values
Out[10]: 
array(['1979-01-01T05:00:00.000000000', '1979-01-01T05:00:00.000000000',
       '1979-01-01T05:00:00.000000000', '1979-01-01T05:00:00.000000000',
       '1979-01-01T05:00:00.000000000', '1979-01-01T05:00:00.000000000',
       '1979-01-01T05:00:00.000000000', '1979-01-01T05:00:00.000000000',
       '1979-01-01T05:00:00.000000000', '1979-01-01T05:00:00.000000000'],
      dtype='datetime64[ns]')
```

selipot commented on clouddrift/datasets.py:
```diff
@@ -157,6 +157,49 @@ def gdp6h(decode_times: bool = True) -> xr.Dataset:
+    max: int | None = None,
+    decode_times: bool = True,
+) -> xr.Dataset:
+    """Returns the NOAA Global Drifter Program (GDP) raw dataset as a ragged array
```
Let's replace here and below "raw" by "un-interpolated"

philippemiron replied:
what about the funcitons' name? Looks like we have `def gdpraw(` for example.

philippemiron replied:
What about `gdp_source` ?  and replace raw dataset with source (uninterpolated) dataset.

kevinsantana11 replied:
do we want `gdp_source` , `gdp_uninterpolated`. We could also use something like `gdp_preprocess`

selipot commented:
If you refer to the [header file](https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/header_buoydata_rawfiles.txt) for the raw files, we are missing some data in the resulting ragged array: "Position Time(MM,DD.dd,YYYY)", "Sensor Time(MM,DD.dd,YYYY)", and "Quality Index". Note that position time and sensor time are not the same (sometimes they are).

kevinsantana11 commented:
> If you refer to the [header file](https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/header_buoydata_rawfiles.txt) for the raw files, we are missing some data in the resulting ragged array: "Position Time(MM,DD.dd,YYYY)", "Sensor Time(MM,DD.dd,YYYY)", and "Quality Index". Note that position time and sensor time are not the same (sometimes they are).

Hmm I could of sworn that I included these. Let me check this over

kevinsantana11 commented:
> If you refer to the [header file](https://www.aoml.noaa.gov/ftp/pub/phod/pub/pazos/data/shane/sst/header_buoydata_rawfiles.txt) for the raw files, we are missing some data in the resulting ragged array: "Position Time(MM,DD.dd,YYYY)", "Sensor Time(MM,DD.dd,YYYY)", and "Quality Index". Note that position time and sensor time are not the same (sometimes they are).

Looks like I forgot to include qualityIndex. Went ahead and added the commit to fix that. I originally had the position and sensor time variables under the names `posObsDatetime` and `sensorObsDatetime`. Considering these are long and not great names for the variables I renamed them to `position_datetime` and `sensor_datetime`.

selipot commented:
@kevinsantana11 ok i think i may have missed the variables `posObsDatetime` and `sensorObsDatetime` previously but now I see the `position_time` and` sensor_time` variables, as well as  `qualityIndex`. However, it looks like there is no attributes for the dataset or the variables as expected from source.py? See screenshot after doing `ds = gdp_source()`

<img width="861" alt="Screenshot 2024-06-24 at 5 52 45 PM" src="https://github.com/Cloud-Drift/clouddrift/assets/7576346/fe916039-4d6f-4cf9-adbb-5a84b4510600">

philippemiron commented on clouddrift/datasets.py:
```diff
@@ -157,6 +158,73 @@ def gdp6h(decode_times: bool = True) -> xr.Dataset:
 
+def gdp_source(
+    tmp_path: str = adapters.gdp_source._TMP_PATH,
+    max: int | None = None,
```
Could we rename this parameter?

philippemiron replied:
@kevinsantana11

philippemiron commented:
Other comments:

- what is `obs_index` in the coordinates?
- drogue metadata says `flag_values: 1,0` but 

```
>>> ds.drogue.values[:10]
array([ 0., 11., 11., 13., 18., 19., 20., 21., 18., 25.], dtype=float32)
```
- some longitudes contains 0, which is a bit suspicious:
```
>>> ds.longitude.values[:10]
array([  0.   ,   0.   , 154.725, 154.725, 154.725, 154.903, 154.903,
       154.903, 154.903, 155.534], dtype=float32)
```
- position_datetime doesn't have metadata
- program_number should probably be integer values
- do we have more info we could add to the metadata of `sensor4`, `sensor5`, `sensor6` ?
- sensor_datetime doesn't have metadata
- start_lat metadata says: `Last good latitude...` instead of `First good latitude...`

selipot commented:
> Other comments:
> 
> * what is `obs_index` in the coordinates?

As I understand it, the index of observations when reading the 4 files consecutively. The first line of the first file should correspond to the minimum value of obs_index, that is zero, and the last line of the last file should be the maximum value of the obs_index (but I don't think it is). obs_index is not sorted because of the parallel execution.

> * drogue metadata says `flag_values: 1,0` but
> 
> ```
> >>> ds.drogue.values[:10]
> array([ 0., 11., 11., 13., 18., 19., 20., 21., 18., 25.], dtype=float32)
> ```

Yes, that is because I think it is engineering values of the drogue presence sensor, not a flag. Metadata is wrong.

>  * some longitudes contains 0, which is a bit suspicious:
> 
> ```
> >>> ds.longitude.values[:10]
> array([  0.   ,   0.   , 154.725, 154.725, 154.725, 154.903, 154.903,
>        154.903, 154.903, 155.534], dtype=float32)
> ```

Not sure that is suspicious. There are crazy values in there including `1000` from rounding a fill value like `999.99`.

>  * position_datetime doesn't have metadata
> * program_number should probably be integer values
> * do we have more info we could add to the metadata of `sensor4`, `sensor5`, `sensor6` ?
> * sensor_datetime doesn't have metadata
> * start_lat metadata says: `Last good latitude...` instead of `First good latitude...`

All of this need to be corrected and investigated.

kevinsantana11 commented:
@selipot are the erroneous values observed in the `drogue` (non 0/1 values) and `longitude` (values not in [-180, 180]) variables likely a result of transmission error?

I think here we should consider either providing information in the metadata regarding the values or replacing them with Nans. I understand we may want to avoid this since we want to perform as little pre-processing of the gdp source data as possible.

@philippemiron the obs index is the row position of the observation if we were to consider all of the data files unified as a single file concatenated in order.

What should we add for the metadata of `sensor4`, `sensor5`, `sensor6`? As I remember it these are there in case there are other sensors installed on the drifter.

I will work on adding the missing metadata for `position_datetime` and  `sensor_datetime`  as well as fixing the start_lat metadata.

selipot commented:
> @selipot are the erroneous values observed in the `drogue` (non 0/1 values) and `longitude` (values not in [-180, 180]) variables likely a result of transmission error?

No, these are values that come out of a tether strain or submergence sensor. These values are then processed by the DAC to determine the time at which the drogue has detached.

> I think here we should consider either providing information in the metadata regarding the values or replacing them with Nans. I understand we may want to avoid this since we want to perform as little pre-processing of the gdp source data as possible.

Nope, the resulting dataset should not transform any of the source data variables.

> @philippemiron the obs index is the row position of the observation if we were to consider all of the data files unified as a single file concatenated in order.
> 
> What should we add for the metadata of `sensor4`, `sensor5`, `sensor6`? As I remember it these are there in case there are other sensors installed on the drifter.

That is correct, these are placeholders in case of additional sensors such as barometers, salinity etc.

> I will work on adding the missing metadata for `position_datetime` and `sensor_datetime` as well as fixing the start_lat metadata.

We need to work with the DAC to write appropriate attributes/metadata for these variables.

kevinsantana11 commented:
@selipot I'm still unsure about how were getting erroneous values for the drogue / longitude variables. Mainly I want to get at the cause of them to provide the information in the metadata (if these values are expected)

selipot commented:
> @selipot I'm still unsure about how were getting erroneous values for the drogue / longitude variables. Mainly I want to get at the cause of them to provide the information in the metadata (if these values are expected)

As I mentioned above the values from the "drogue" column come out of a tether strain or submergence sensor. These values are not 0 or 1 (for the drogue status) but whatever the sensor outputs in whatever unit which depends on the sensor type and manufacturer. These values are then processed by the DAC to determine the time at which the drogue has detached. I am unsure what the metadata should say but we can think about it.

As for longitude values being zero, myriads of reasons can lead to erroneous values. It is way beyond the scope of this PR to investigate this.

philippemiron commented:
Last thing, I might be missing it in the pr but in the other GDP adapters we have:

```
return gdp.order_by_date(
        gdp_metadata, [int(f.split("_")[-1].removesuffix(".nc")) for f in filelist]
    )
```

to order the trajectories in the ragged array by their starting time.

I think we should have something like this for consistency and keep the order of the ragged array reproducible.

kevinsantana11 commented:
> Last thing, I might be missing it in the pr but in the other GDP adapters we have:
> 
> ```
> return gdp.order_by_date(
>         gdp_metadata, [int(f.split("_")[-1].removesuffix(".nc")) for f in filelist]
>     )
> ```
> 
> to order the trajectories in the ragged array by their starting time.
> 
> I think we should have something like this for consistency and keep the order of the ragged array reproducible.

Due to the distributed nature, were going to have to do this in a different way if we want to order the drifter segments in ascending order using the start_date but I think we may also need to look into the `gdp.order_by_date`  function. Currently the body looks like:

```python
return df.ID[np.where(np.in1d(df.ID, idx))[0]].values  # type: ignore
```

philippemiron commented:
Not sure this would work, but once you have to fully form `xr.Dataset`, should be possible to just sort by time. We can do this in another pr.

selipot commented:
I just tested this and it seems to be working well and consistently. I think the drogue variable attributes still need correction. I would need a little bit of time to check the output.

selipot commented:
What is holding this one ?

kevinsantana11 commented:
> What is holding this one ?
I'm mainly looking into cleaning this up and adding any missing documentation. I looked into the filtering and 100,000 records are filtered out due to their values. I documented the [filters in the code](https://github.com/kevinsantana11/clouddrift/blob/raw-gdp-files-adapter/clouddrift/adapters/gdp/source.py#L360-L375)

There's currently a second filtering step that occurs where we ignore drifter observations associated to drifters not found in the metadata file. Do we want to ignore these as well or do we want to fill it with dummy data?

total_obs = 332_711_844
filtered_bad_data_obs = 100_000
post_filtering_obs = 332_274_093

total_filtered_obs = 437_751

kevinsantana11 commented:
@selipot I went ahead and updated this code to not drop the observatons associated to drifters missing metadata. Instead for those I have provided an option where you can decide to populate those with fill values or have them ignored as before.

selipot commented:
> @selipot I went ahead and updated this code to not drop the observatons associated to drifters missing metadata. Instead for those I have provided an option where you can decide to populate those with fill values or have them ignored as before.

Please have a look at section 2.5.1 of the CF conventions [Missing data, valid and actual range of data](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html). We want to make sure we are adding the proper attributes including the `_FillValue` attribute.

philippemiron commented:
I think xarray takes care of setting appropriate `_FillValue` when writing out to netCDF. Is there something that is wrong?

I think we should try to merge this PR, it is modifying too many files to be sitting around for that long.

> > @selipot I went ahead and updated this code to not drop the observatons associated to drifters missing metadata. Instead for those I have provided an option where you can decide to populate those with fill values or have them ignored as before.
> 
> Please have a look at section 2.5.1 of the CF conventions [Missing data, valid and actual range of data](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html). We want to make sure we are adding the proper attributes including the `_FillValue` attribute.

kevinsantana11 commented:
@philippemiron I could also separate the changes that are unrelated to the adapter. I think that will also make things cleaner.

philippemiron commented:
Let's see what's left to do; I don't think it's worth taking the time to reorganize everything.

kevinsantana11 commented:
gg

> Let's see what's left to do; I don't think it's worth taking the time to reorganize everything.

Only seeing this now after spending the time re-organizing :sweat_smile: I'm still glad I separated things out individually as the git history and commits will be easier to read if needed but it definetly took a good bit of time.

kevinsantana11 commented:
Updated the code to leverage dask instead of using our own custom parallelization code. The change is still WIP

kevinsantana11 commented:
I Separated the change to use dask over to another [PR](https://github.com/Cloud-Drift/clouddrift/pull/508). This PR should be good to review and potentially merge in @selipot
